core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -1539,4 +1540,86 @@ public TaskDef getTaskDefinition(Task task) {
+
+        // On Subworkflow complete or terminate..
+        if (subWorkflow.getStatus().isTerminal()) {
+            // If parent is FAILED and the sub workflow task in parent is FAILED, we want to resume them
+            if (parentWorkflow.getStatus().equals(WorkflowStatus.FAILED)) {
+                String errorMsg = String.format("Subworkflow: %s is successful, but parent workflow: %s is in FAILED state.",
+                        subWorkflow.getWorkflowId(), parentWorkflow.getWorkflowId());
+                LOGGER.warn(errorMsg);
+                throw new IllegalStateException(errorMsg);
+            } else if (subWorkflowTask.getStatus().equals(IN_PROGRESS)) {
+                LOGGER.debug("Subworkflow: {} is {}, updating parent workflow: {}",
+                        subWorkflow.getWorkflowId(), subWorkflow.getStatus().name(), parentWorkflow.getWorkflowId());
+                SubWorkflow subWorkflowSystemTask = new SubWorkflow();
+                subWorkflowSystemTask.execute(subWorkflow, subWorkflowTask, this);
+                // Keep Subworkflow task's data consistent with Subworkflow's.
+                if (subWorkflowTask.getStatus().isTerminal() && subWorkflowTask.getExternalOutputPayloadStoragePath() != null && !subWorkflowTask.getOutputData().isEmpty()) {
+                    Map<String, Object> parentWorkflowTaskOutputData = subWorkflowTask.getOutputData();
+                    deciderService.populateTaskData(subWorkflowTask);
+                    subWorkflowTask.getOutputData().putAll(parentWorkflowTaskOutputData);
+                    deciderService.externalizeTaskData(subWorkflowTask);
+                }
+                return true;
+            } else {
+                LOGGER.warn("Unable to do something with parent workflow: {} in status: {}, and subworkflow: {} in status: {}",
+                        parentWorkflow.getWorkflowId(), parentWorkflow.getStatus().name(),
+                        subWorkflow.getWorkflowId(), subWorkflow.getStatus().name());
+            }
+        } else {
+            // On workflow retry or restart..
+            if (StringUtils.isBlank(parentDef.getFailureWorkflow()) && parentWorkflow.getStatus().isTerminal() && subWorkflowTask.getStatus().isTerminal()) {
+                LOGGER.debug("Subworkflow: {} is {}, resetting failed parent workflow: {}, and Subworkflow task: {} status to IN_PROGRESS",
+                        subWorkflow.getWorkflowId(), subWorkflow.getStatus().name(), parentWorkflow.getWorkflowId(), subWorkflow.getParentWorkflowTaskId());
+                subWorkflowTask.setStatus(IN_PROGRESS);
+                executionDAOFacade.updateTask(subWorkflowTask);
+                parentWorkflow.setStatus(WorkflowStatus.RUNNING);
+                executionDAOFacade.updateWorkflow(parentWorkflow);
+                return true;
+            } else if (parentWorkflow.getStatus().equals(WorkflowStatus.RUNNING)) {
+                if (subWorkflowTask.getStatus().isTerminal()) {
```

Comment:
What is the scenario when this would be invoked?

Reply:
Ideally, it shouldn't. But, until we have recovery, if the workflow were to run into inconsistent state because of previous transient errors, this helps stops it from further continuing the inconsistency.